torchao/quantization/marlin_qqq/__init__.py

```diff
@@ -0,0 +1,373 @@
+import torch
+
+from torchao.quantization.marlin_qqq.utils import (
+    marlin_permute_weights,
+    get_pack_factor,
+    get_qqq_scale_perms,
+    get_qqq_weight_perm,
+    get_qqq_scale_reverse_perms,
+    get_qqq_weight_reverse_perm,
+    reverse_marlin_permute_weights,
+    const,
+)
+
+__all__ = [
+    "marlin_qqq_workspace",
+    "qqq_quantize_weights",
+    "pack_to_marlin_qqq",
+    "unpack_from_marlin_qqq",
+]
+
+
+def marlin_qqq_workspace(
+    out_features: int,
+    min_thread_n: int = const.MIN_THREAD_N,
+    max_parallel: int = const.MAX_PARALLEL,
+) -> torch.Tensor:
+    """Creates a workspace for marlin qqq. The workspace is used to coordinate the locks
+    during the execution of the kernel.
+
+    Args:
+        out_features (int): The number of output features.
+        min_thread_n (int, optional): The minimum number of threads per block. Defaults to `MARLIN_QQQ_MIN_THREAD_N`.
+        max_parallel (int, optional): The maximum number of parallel threads. Defaults to `MARLIN_QQQ_MAX_PARALLEL`.
+    Returns:
+        torch.Tensor: The workspace tensor fully initialized with zeros.
+    """
+    assert (
+        out_features % min_thread_n == 0
+    ), f"out_features = {out_features}, min_thread_n = {min_thread_n}"
+    max_workspace_size = (out_features // min_thread_n) * max_parallel
+    return torch.zeros(max_workspace_size, dtype=torch.int, device="cuda")
+
+
+# QQQ employs different quant schemes for per-group and
+# per-channel quantization.
+def qqq_quantize_weights(
+    w: torch.Tensor, num_bits: int, group_size: int
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Quantize the weight.
+
+    Args:
+        w (torch.Tensor): The original weight.
+        num_bits (int): The number of bits for quantization.
+        group_size (int): The group size of quantization.
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        The dequantized weight, the quantized weight, the per-group quantization scale,
+        and the per-channel quantization scale.
+    """
+    orig_device = w.device
+    size_k, size_n = w.shape
+
+    assert w.is_floating_point(), "w must be float"
+    assert num_bits in const.SUPPORTED_NUM_BITS, f"Unsupported num_bits = {num_bits}"
+    assert group_size in const.SUPPORTED_GROUP_SIZES + [
+        size_k
+    ], f"Unsupported groupsize = {group_size}"
+
+    if group_size == -1:
+        group_size = size_k
+    assert group_size <= size_k
+
+    if group_size < size_k:
+        # Reshape to [groupsize, -1]
+        w = w.reshape((-1, group_size, size_n))
+        w = w.permute(1, 0, 2)
+        w = w.reshape((group_size, -1))
+
+        max_q_val = 2**num_bits - 1
+        half_q_val = (max_q_val + 1) // 2
+
+        # Compute scale for each group
+        s_group = torch.max(torch.abs(w), 0, keepdim=True)[0]
+        s_group *= 2 / max_q_val  # 2 => symmetric
+
+        # Quantize
+        q_w = torch.round(w / s_group).int()
```

Comment:
cc @jerryzh168  I think we should make this block-size to -> max_abs a utility

Reply:
it's just https://github.com/pytorch/ao/blob/e41ca4ee41f5f1fe16c59e00cffb4dd33d25e56d/torchao/quantization/quant_primitives.py#L773-L777 I think it's possible to reuse already